patches/server/0950-Paper-Item-API.patch

```diff
@@ -0,0 +1,4333 @@
++import net.kyori.adventure.text.event.HoverEvent;
++import net.minecraft.world.item.Item;
++import org.bukkit.Material;
++import org.bukkit.craftbukkit.util.CraftMagicNumbers;
++import org.bukkit.inventory.meta.ItemMeta;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.function.UnaryOperator;
++
++public class VanillaItemStack implements ItemStack {
++
++    private final net.minecraft.world.item.ItemStack itemStack;
++
++    public VanillaItemStack(@NotNull net.minecraft.world.item.ItemStack vanilla) {
++        this.itemStack = vanilla;
++    }
++
++    @Override
++    public @NotNull Material getMaterial() {
++        return CraftMagicNumbers.getMaterial(this.itemStack.getItem());
++    }
++
++    @Override
++    public void setMaterial(@NotNull Material material) {
++        this.itemStack.setItem(CraftMagicNumbers.getItem(material));
++    }
++
++    @Override
++    public int getSize() {
++        return this.itemStack.getCount();
++    }
++
++    @Override
++    public void setSize(int size) {
++        this.itemStack.setCount(size);
++    }
++
++    @Override
++    public @Nullable ItemMeta getMeta() {
++        return null;
++    }
++
++    @Override
++    public void setMeta(@Nullable ItemMeta meta) {
++
++    }
++
++    @Override
++    public @NotNull HoverEvent<HoverEvent.ShowItem> asHoverEvent(@NotNull UnaryOperator<HoverEvent.ShowItem> op) {
++        Key key = CraftMagicNumbers.getMaterial(this.itemStack.getItem()).getKey();
++
++        return HoverEvent.showItem(op.apply(HoverEvent.ShowItem.of(key, this.itemStack.getCount(), PaperAdventure.asBinaryTagHolder(this.itemStack.getTag()))));
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/item/MetaPropertySkeletons.java b/src/main/java/io/papermc/paper/item/MetaPropertySkeletons.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..ed92fbdf46183ce5a84e2f57cef23e6d7cab356f
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/item/MetaPropertySkeletons.java
+@@ -0,0 +1,59 @@
++package io.papermc.paper.item;
++
++import io.papermc.paper.property.PropertyHolder;
++import io.papermc.paper.property.PropertyKey;
++import io.papermc.paper.property.value.Value;
++import org.bukkit.potion.PotionData;
++import org.bukkit.potion.PotionType;
++
++import java.util.List;
++
++import static io.papermc.paper.inventory.item.properties.ItemProperties.*;
++
++public interface MetaPropertySkeletons {
++
++
++
++    List<PropertyKey<?>> CRAFT_LEATHER_ARMOR = List.of(
++        COLOR
++    );
++
++    List<PropertyKey<?>> CRAFT_REPAIRABLE = List.of(
++        REPAIR_COST
++    );
++
++    // CraftMetaPotion
++//    PropertyHolder.Immutable DEFAULT_STATE = PropertyHolder.Immutable.builder()
++//        .of(DEFAULT_POTION_TYPE, Value.of(DEFAULT_POTION_TYPE, new PotionData(PotionType.UNCRAFTABLE, false, false)))
++//        .build();
++
++    List<PropertyKey<?>> CRAFT_META_POTION = List.of(
```

Comment:
I see that this is for validating which property can be used on which type of ItemStack.
Do we have a possibility to populate these kind of List dynamically using server implementation?

The dirty idea is to generate a default itemstack for each Materials once on startup, iterate over the nbt keys and check if a PropertyKey is applicable for that stack.

Reply:
Not exactly, that was for something that was outdated. Generally, every property key can be applied to any stack. There is no "compatibility" with material types.